### PR TITLE
[HOPSWORKS-1114] reset password

### PR DESCRIPTION
**Pull request by ErmiasG**
## Make sure there is no duplicate PR for this issue

* **Please check if the PR meets the following requirements**
- [x] Adds tests for the submitted changes (for bug fixes & features)
- [ ] Passes the tests
- [x] HOPSWORKS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit

* **Post a link to the associated JIRA issue**

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

**Comment by ErmiasG** [comment]
logicalclocks/hops-testing#441

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/user/UsersController.java [hunk 1]
```diff
@@ -288,40 +282,97 @@ public void addOrg(Users user) {
...
+    authController.resetValidationKey(user);
+    //Send verification
+    String subject = UserAccountsEmailMessages.ACCOUNT_QR_RESET;
+    String msg = UserAccountsEmailMessages.buildQRResetMessage();
+    emailBean.sendEmail(user.getEmail(), Message.RecipientType.TO, subject, msg);
+    return Base64.encodeBase64(qrCode);
+  }
+  
+  private byte[] recoverQRCode(Users user, HttpServletRequest req) {
+    String random = SecurityUtils.getRandomPassword(Settings.PASSWORD_MIN_LENGTH);
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/user/UsersController.java]: Why not use `getSecureRandom` here?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/user/AuthController.java [hunk 2]
```diff
@@ -463,6 +604,11 @@ public void registerFalseLogin(Users user, HttpServletRequest req) {
       userFacade.update(user);
     }
   }
+  
+  private boolean isUserAgent(Users user) {
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/user/AuthController.java]: We already have `isUserInRole` method. We should reuse that one.

File: hopsworks-rest-utils/src/main/java/io/hops/hopsworks/restutils/RESTCodes.java [hunk 3]
```diff
@@ -835,7 +835,8 @@ public int getRange() {
     ACCESS_CONTROL(47, "Client not authorized for this invocation.", Response.Status.FORBIDDEN),
     SECRET_EMPTY(48, "Secret is empty", Response.Status.NOT_FOUND),
     SECRET_EXISTS(49, "Same Secret already exists", Response.Status.CONFLICT),
-    SECRET_ENCRYPTION_ERROR(50, "Error encrypting/decrypting Secret", Response.Status.INTERNAL_SERVER_ERROR);
+    SECRET_ENCRYPTION_ERROR(50, "Error encrypting/decrypting Secret", Response.Status.INTERNAL_SERVER_ERROR),
+    ACCOUNT_NOT_ACTIVE(51, "This account is not active", Response.Status.BAD_REQUEST);
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-rest-utils/src/main/java/io/hops/hopsworks/restutils/RESTCodes.java]: yes but not the same Response.Status. 4 will redirect you to login page but 51 is a bad request

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/util/Settings.java [hunk 4]
```diff
@@ -2315,7 +2316,12 @@ public synchronized String getHopsExamplesFeaturestoreFilename() {
     return "hops-examples-featurestore-" + HOPS_EXAMPLES_VERSION + ".jar";
   }
 
-  private String VERIFICATION_PATH = "hopsworks-api/api/auth/verify";
+  private String VERIFICATION_PATH = "/hopsworks-admin/security/validate_account.xhtml";
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/util/Settings.java]: did not want this pr to touch too many parts. but I did prepare for that change

File: hopsworks-jwt/src/main/java/io/hops/hopsworks/jwt/JWTController.java [hunk 5]
```diff
@@ -714,6 +714,7 @@ public boolean markOldSigningKeys() {
     final Calendar cal = Calendar.getInstance();
     cal.add(Calendar.DATE, -Constants.ONE_TIME_JWT_SIGNING_KEY_ROTATION_DAYS);
     if (jwtSigningKey != null && jwtSigningKey.getCreatedOn().before(cal.getTime())) {
+      removeMarkedKeys();//remove if there is an old marked but not deleted.
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-jwt/src/main/java/io/hops/hopsworks/jwt/JWTController.java]: I want to go back to the commented code but was not sure why it was changed and what to test

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/user/UsersController.java (hunk 1 above)
Comment (ErmiasG) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/user/UsersController.java]: This code is just moved from https://github.com/logicalclocks/hopsworks/pull/279/files/127647f65e3d2e20a1b7f3a28ebb107b8648b32b#diff-7b18d017a63053a6feb7d4ed10f37140L199

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/user/AuthController.java (hunk 2 above)
Comment (ErmiasG) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/user/AuthController.java]: just a private method, I think it is more readable this way

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/user/UsersController.java [hunk 6]
```diff
@@ -121,34 +121,29 @@
...
-    NoSuchAlgorithmException, UserException {
+  public byte[] registerUser(UserDTO newUser, HttpServletRequest req) throws NoSuchAlgorithmException, UserException {
     userValidator.isValidNewUser(newUser);
     Users user = createNewUser(newUser, UserAccountStatus.NEW_MOBILE_ACCOUNT, UserAccountType.M_ACCOUNT_TYPE);
     addAddress(user);
     addOrg(user);
-    //to privent sending email for test user emails
+    //to prevent sending email for test user emails
     try {
       if (!newUser.isTestUser()) {
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/user/UsersController.java]: it is just in the request. To not send email if it is test

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/user/UsersController.java [hunk 7]
```diff
@@ -121,34 +121,29 @@
...
         // Notify user about the request if not test user.
-        emailBean.sendEmail(newUser.getEmail(), RecipientType.TO, UserAccountsEmailMessages.ACCOUNT_REQUEST_SUBJECT,
-            UserAccountsEmailMessages.buildMobileRequestMessage(FormatUtils.getUserURL(req), user.getUsername() + user.
-                getValidationKey()));
+        authController.sendEmailValidationKey(user, user.getValidationKey(), req);
       }
       // Only register the user if i can send the email
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/user/UsersController.java]: yes but we can get spammed with fake emails. And we do not want that!

File: hopsworks-IT/pom.xml [hunk 8]
```diff
@@ -60,10 +60,13 @@
     <commons.lang3.version>3.8.1</commons.lang3.version>
     <plugin.failsafe.version>2.16</plugin.failsafe.version>
     <!-- Glassfish conf -->
+    <!--suppress UnresolvedMavenProperty -->
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-IT/pom.xml]: yes but it was annoying so I kept it. Is that a problem?

[the remark this answers is not in the document]
Comment (tkakantousis) [reply hopsworks-IT/pom.xml]: No

File: hopsworks-IT/src/test/java/io/hops/hopsworks/util/DBHelper.java [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+public class DBHelper {
+  private static final String DB_HOST_ENV = "DB_HOST";
+  private static final String DB_PORT_ENV = "DB_PORT";
+  private static final String DEFAULT_PORT = "3306";
+  private static final String DEFAULT_HOST = "127.0.0.1";
+  private static final String JDBC_DRIVER = "com.mysql.jdbc.Driver";
+  private static final String DB_URL = "jdbc:mysql://%%s%%/hopsworks";
+  private static final String USER = "kthfs";
+  private static final String PASSWORD = "kthfs";
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply hopsworks-IT/src/test/java/io/hops/hopsworks/util/DBHelper.java]: You mean the password?

[the remark this answers is not in the document]
Comment (tkakantousis) [reply hopsworks-IT/src/test/java/io/hops/hopsworks/util/DBHelper.java]: yes

**Pull request merged**